File: android/ac-copilot-extensions/src/main/kotlin/com/microsoft/adaptivecards/copilot/ChainOfThoughtView.kt
package com.microsoft.adaptivecards.copilot

import androidx.compose.animation.*
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Psychology
import coil.compose.AsyncImage

/**
 * Jetpack Compose view for rendering Chain of Thought UX.
 *
 * Ported from production Teams-AdaptiveCards-Mobile SDK.
 * Shows the reasoning steps Copilot goes through while processing a request.
 */
@Composable
fun ChainOfThoughtView(
    data: ChainOfThoughtData,
    modifier: Modifier = Modifier,
    onHeightChange: (() -> Unit)? = null
) {
    var expandedSteps by remember { mutableStateOf(setOf(0)) }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .shadow(4.dp, RoundedCornerShape(12.dp))
            .clip(RoundedCornerShape(12.dp))
            .background(MaterialTheme.colorScheme.surface)
            .padding(16.dp)
    ) {
        // Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(
                    MaterialTheme.colorScheme.surfaceVariant,
                    RoundedCornerShape(8.dp)
                )
                .padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = Icons.Filled.Psychology,
                contentDescription = "Thinking",
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(16.dp)
            )
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = data.state,
                style = MaterialTheme.typography.titleSmall,
                fontWeight = FontWeight.Medium
            )
        }

        Spacer(modifier = Modifier.height(8.dp))

        // Entries
        data.entries.forEachIndexed { index, entry ->
            ChainOfThoughtEntryView(
                entry = entry,
                isCompleted = data.isDone || index < data.entries.size - 1,
                isLast = index == data.entries.size - 1,
                isExpanded = expandedSteps.contains(index),
                onToggleExpanded = {
                    onHeightChange?.invoke()
                    expandedSteps = if (expandedSteps.contains(index)) {
                        expandedSteps - index
                    } else {
                        expandedSteps + index
                    }
                }
            )
        }
    }
}

@Composable
private fun ChainOfThoughtEntryView(
    entry: ChainOfThoughtEntry,
    isCompleted: Boolean,
    isLast: Boolean,
    isExpanded: Boolean,
    onToggleExpanded: () -> Unit
) {
    Column(modifier = Modifier.fillMaxWidth()) {
        // Header row
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .clickable { onToggleExpanded() }
                .padding(vertical = 8.dp),
            verticalAlignment = Alignment.Top
        ) {
            // Status indicator
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier.width(12.dp)
            ) {
                Box(
                    modifier = Modifier
                        .size(12.dp)
                        .clip(CircleShape)
                        .background(if (isCompleted) Color(0xFF4CAF50) else Color(0xFFFF9800)),
                    contentAlignment = Alignment.Center
                ) {
                    if (isCompleted) {
                        Text("✓", fontSize = 8.sp, color = Color.White, fontWeight = FontWeight.Bold)
                    }
                }

                if (!isLast) {
                    Spacer(modifier = Modifier.height(4.dp))
                    Box(
                        modifier = Modifier
                            .width(2.dp)
                            .height(24.dp)
                            .background(Color.Gray.copy(alpha = 0.3f))
                    )
                }
            }

            Spacer(modifier = Modifier.width(12.dp))

            // Content
            Column(modifier = Modifier.weight(1f)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = entry.header,
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.SemiBold,
                        modifier = Modifier.weight(1f)
                    )

                    // App info
                    entry.appInfo?.let { appInfo ->
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            modifier = Modifier.padding(start = 8.dp)
                        ) {
                            AsyncImage(
                                model = appInfo.icon,
                                contentDescription = appInfo.name,
                                modifier = Modifier.size(16.dp),
                                contentScale = ContentScale.Fit
                            )
                            Spacer(modifier = Modifier.width(4.dp))
                            Text(
                                text = appInfo.name,
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }

                    // Chevron
                    Text(
                        text = if (isExpanded) "▲" else "▼",
                        fontSize = 12.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(start = 8.dp)
                    )
                }
            }
        }

        // Expanded content
        AnimatedVisibility(visible = isExpanded) {
            Row(modifier = Modifier.padding(start = 24.dp, bottom = 8.dp)) {
                Text(
                    text = entry.content,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

File: ios/Sources/ACCopilotExtensions/ChainOfThoughtView.swift
import SwiftUI

// MARK: - ChainOfThoughtView

/// SwiftUI view for rendering Chain of Thought UX.
///
/// Ported from production Teams-AdaptiveCards-Mobile SDK.
/// Shows the reasoning steps Copilot goes through while processing a request,
/// with expandable entries, status indicators, and animated transitions.
@available(iOS 15.0, *)
public struct ChainOfThoughtView: View {
    public let data: ChainOfThoughtData
    @State private var expandedSteps = Set<Int>()
    public var onHeightChange: (() -> Void)?

    public init(data: ChainOfThoughtData, onHeightChange: (() -> Void)? = nil) {
        self.data = data
        self.onHeightChange = onHeightChange
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Header with state
            HStack {
                Image(systemName: "brain.head.profile")
                    .foregroundColor(.blue)
                    .font(.system(size: 16))

                Text(data.state)
                    .font(.headline)
                    .fontWeight(.medium)
                    .foregroundColor(.primary)

                Spacer()
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            #if canImport(UIKit)
            .background(Color(UIColor.systemGray6))
            #else
            .background(Color.gray.opacity(0.12))
            #endif
            .cornerRadius(8)

            // Chain of thought entries
            VStack(alignment: .leading, spacing: 0) {
                ForEach(Array(data.entries.enumerated()), id: \.offset) { index, entry in
                    ChainOfThoughtEntryView(
                        entry: entry,
                        stepIndex: index,
                        isCompleted: data.isDone || index < data.entries.count - 1,
                        isLast: index == data.entries.count - 1,
                        isExpanded: expandedSteps.contains(index),
                        onToggleExpanded: {
                            onHeightChange?()
                            withAnimation(.easeInOut(duration: 0.3)) {
                                if expandedSteps.contains(index) {
                                    expandedSteps.remove(index)
                                } else {
                                    expandedSteps.insert(index)
                                }
                            }
                        }
                    )
                }
            }
            .padding(.horizontal, 8)
            .padding(.top, 8)
        }
        .padding(16)
        .background(Color(white: 1.0, opacity: 1.0))
        .cornerRadius(12)
        .shadow(color: .black.opacity(0.1), radius: 4, x: 0, y: 2)
        .frame(maxWidth: .infinity)
        .fixedSize(horizontal: false, vertical: true)
        .onAppear {
            if !data.entries.isEmpty {
                expandedSteps.insert(0)
            }
        }
    }
}

// MARK: - ChainOfThoughtEntryView

@available(iOS 15.0, *)
public struct ChainOfThoughtEntryView: View {
    public let entry: ChainOfThoughtEntry
    public let stepIndex: Int
    public let isCompleted: Bool
    public let isLast: Bool
    public let isExpanded: Bool
    public let onToggleExpanded: () -> Void

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Step header
            Button(action: onToggleExpanded) {
                HStack(alignment: .top, spacing: 12) {
                    // Status indicator with connecting line
                    VStack(spacing: 0) {
                        ZStack {
                            Circle()
                                .fill(isCompleted ? Color.green : Color.orange)
                                .frame(width: 12, height: 12)

                            if isCompleted {
                                Image(systemName: "checkmark")
                                    .font(.system(size: 8, weight: .bold))
                                    .foregroundColor(.white)
                            }
                        }

                        if !isLast {
                            Rectangle()
                                .fill(Color.gray.opacity(0.3))
                                .frame(width: 2, height: 24)
                                .padding(.top, 4)
                        }
                    }
                    .frame(width: 12)

                    // Header content
                    VStack(alignment: .leading, spacing: 4) {
                        HStack(alignment: .center) {
                            Text(entry.header)
                                .font(.subheadline)
                                .fontWeight(.semibold)
                                .lineLimit(nil)
                                .multilineTextAlignment(.leading)
                                .foregroundColor(.primary)
                                .fixedSize(horizontal: false, vertical: true)

                            Spacer()

                            // App info
                            if let appInfo = entry.appInfo {
                                HStack(spacing: 4) {
                                    AsyncImage(url: URL(string: appInfo.icon)) { image in
                                        image.resizable().aspectRatio(contentMode: .fit)
                                    } placeholder: {
                                        RoundedRectangle(cornerRadius: 4)
                                            .fill(Color.gray.opacity(0.3))
                                    }
                                    .frame(width: 16, height: 16)

                                    Text(appInfo.name)
                                        .font(.caption)
                                        .foregroundColor(.secondary)
                                }
                            }

                            // Chevron
                            Image(systemName: isExpanded ? "chevron.up" : "chevron.down")
                                .foregroundColor(.gray)
                                .font(.system(size: 12))
                        }
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
            .buttonStyle(PlainButtonStyle())
            .padding(.vertical, 8)

            // Expanded content
            if isExpanded {
                HStack(alignment: .top, spacing: 12) {
                    VStack(spacing: 0) {
                        Spacer().frame(height: 12)
                        if !isLast {
                            Rectangle()
                                .fill(Color.gray.opacity(0.3))
                                .frame(width: 2)
                        }
                    }
                    .frame(width: 12)

                    Text(entry.content)
                        .font(.caption)
                        .foregroundColor(.secondary)
                        .lineLimit(nil)
                        .fixedSize(horizontal: false, vertical: true)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(.bottom, 8)
                }
                .transition(.asymmetric(
                    insertion: .opacity.combined(with: .move(edge: .top)),
                    removal: .opacity.combined(with: .move(edge: .top))
                ))
            }
        }
        .frame(maxWidth: .infinity)
    }
}
